3_Download_HyP3_Data.ipynb

```
@@ -0,0 +1,681 @@
+    "    print(\"Success: Found at least 1 Incidence Angle Map.\")\n",
+    "else:\n",
+    "    raise FileNotFoundError(\"Failed to find at least 1 Incidence Angle Map. \\\n",
+    "    \\nYou will not be able to successfully run a MintPy time-series unless your reorder your HyP3 project \\\n",
+    "with 'Include Inc. Angle Map' option selected.\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "48e233a7-32fe-4404-a536-a4b2827460b8",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "## 4. Subset the Stack\n",
+    "\n",
+    "**Project all tiffs to Predominant UTM**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5d2d85ad-a4c3-48ae-b0ab-3e2c307464c4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dems.sort()\n",
+    "burst = 'BURST' in product_type\n",
+    "if not burst:\n",
+    "    epsgs = util.get_projections(dems)\n",
+    "    predominant_epsg = None if len(epsgs) == 1 else max(epsgs, key=epsgs.get)\n",
+    "    \n",
+    "    if predominant_epsg:\n",
+    "        for pth in dems:\n",
+    "            src_SRS = util.get_projection(str(pth))\n",
+    "            res = util.get_res(pth)\n",
+    "            if src_SRS != predominant_epsg:\n",
+    "                res = util.get_res(pth)\n",
+    "                no_data_val = util.get_no_data_val(pth)\n",
+    "        \n",
+    "                temp = pth.parent/f\"temp_{pth.stem}.tif\"\n",
+    "                pth.rename(temp)\n",
+    "    \n",
+    "                warp_options = {\n",
+    "                    \"dstSRS\":f\"EPSG:{predominant_epsg}\", \"srcSRS\":f\"EPSG:{src_SRS}\",\n",
+    "                    \"targetAlignedPixels\":True,\n",
+    "                    \"xRes\":res, \"yRes\":res,\n",
+    "                    \"dstNodata\": no_data_val\n",
+    "                }\n",
+    "                gdal.Warp(str(pth), str(temp), **warp_options)\n",
+    "                temp.unlink()\n"
+   ]
```

We could remove some indents from this cell by inverting the conditionals and returning early:
```python
if burst:
    break # How do you exit a cell?

epsgs = util.get_projections(dems)
predominant_epsg = None if len(epsgs) == 1 else max(epsgs, key=epsgs.get)
    
if predominant_epsg is None:
    break

for pth in dems:
    src_SRS = util.get_projection(str(pth))
    res = util.get_res(pth)

    if src_SRS == predominant_epsg:
        continue

    res = util.get_res(pth)
    no_data_val = util.get_no_data_val(pth)
        
    temp = pth.parent/f"temp_{pth.stem}.tif"
    pth.rename(temp)
    
    warp_options = {
        "dstSRS":f"EPSG:{predominant_epsg}", "srcSRS":f"EPSG:{src_SRS}",
        "targetAlignedPixels":True,
        "xRes":res, "yRes":res,
        "dstNodata": no_data_val
    }
    gdal.Warp(str(pth), str(temp), **warp_options)
    temp.unlink()
```

---

You can use `sys.exit("reasons")` but it will produce an exception and traceback. It should silently skip reprojection if the conditions aren't met.

I don't know of a native Jupyter way to stop code cell execution programmatically from the code cell. We could probably write a custom magic command that uses jupyter-server to do it. That could be fun.

---

You'd have to have the custom magic installed, maybe before JupyterLab was running, in order to use it. So maybe not the best idea. People should be able to build the conda environment in JupyterLab without restarting it.

---

After looking around a bit, it seems like there isn't a good way to apply this pattern to a notebook cell -- I'd be fine with leaving it as is.